### ISSUE TITLE
Laplace entropy bug is still there

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#116]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/116) by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Check out the following: see where the answer diverges in the second loop?

Turn this into a test and fix it!

```python
In [19]: np.random.seed(100)
    ...: x = mass.MnKAlphaDistribution().rvs(size=1000)
    ...: gain = np.linspace(-.01, .01, 21)
    ...: for p in gain:
    ...:     xg = np.exp(p)*x
    ...:     D = mass.mathstat.entropy.laplace_cross_entropy(xg, x, w=3.0, approx_mode="exact")
    ...:     print D
    ...:     
8.63307826415
8.66526379063
8.32071111374
8.10993392377
7.8423729819
7.52141334287
6.76855794821
5.74840926501
4.69108269918
3.98149506165
3.60715593668
4.02620368536
4.85014239611
5.76470212279
6.59987289814
7.63083288661
8.83701135305
9.7491336354
9.81198817892
9.53953319745
9.50649050668

In [20]: np.random.seed(100)
    ...: x = mass.MnKAlphaDistribution().rvs(size=1000)
    ...: gain = np.linspace(-.01, .01, 21)
    ...: for p in gain:
    ...:     xg = np.exp(p)*x
    ...:     D = mass.mathstat.entropy.laplace_cross_entropy(xg, x, w=1.0, approx_mode="exact")
    ...:     print D
    ...:     
133097622.422
315484.825279
752.203691496
10.832044513
8.1755715008
8.68573186985
7.73485004692
6.66511361704
4.98003576846
4.28756953346
3.30046184799
10422227644.3
6.9373824952e+11
2250243934.61
8327927902.84
19827063.9629
46937.1221928
125.272609245
13.9418313638
11.7592955581
11.679981645
```

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

This appears to be totally fundamental to the notion of the cross-entropy (or KL divergence) whenever the "true" distribution is determined from a vector of sampled values. The problem is that the cross-entropy of measured distribution p given true distribution q will diverge wherever q->0. But if q is KDE-estimated from a set of measurements, then there are places where q gets incredibly small even though the true probability is not that small--we just happened not to sample any data near there.

We can try using a minimum floor on the probability q(x), or try using very heavy-tailed distributions (e.g., Cauchy), but both wipe out the convenience of using the Laplace distribution.

Instead, let q(x) be estimated from sampled values as a sum of top-hat (piecewise-constant) probabilities. We can add to that some q_min floor as we wish without changing the math of the definite integrals. This seems to change measured entropies by ~0.1 nats or less.

Experiment with this.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Change the cross-entropy computation so that the kernel for KDE-
estimation of the true distribution’s kernel is piecewise constant.
This makes it easier to force a non-zero value of the q(x) in
addition to the kernel-estimated values.

Addresses #116.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Clarifying comments about the new cross-entropy.

Also update the doc string, improve variable names, 
fix a tiny bug in computing cross-entropy of the semi-infinite
intervals. Update tests to match the new computation. 
Fixes #116.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.6 (automated comment)
